components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go

```diff
@@ -437,23 +524,63 @@ func (p *PullConfigsImpl) fetchInfraConfig() (*ExistingInfraConfigToml, error) {
-	for _, ele := range a2ConfigMap {
-		if ele.Global.V1.External.Opensearch != nil &&
-			ele.Global.V1.External.Opensearch.Auth != nil &&
-			ele.Global.V1.External.Opensearch.Auth.AwsOs != nil {
-			return setExternalOpensearchDetails(ele.Global.V1.External.Opensearch.Nodes[0].Value,
-				ele.Global.V1.External.Opensearch.Auth.AwsOs.Username.Value,
-				ele.Global.V1.External.Opensearch.Auth.AwsOs.Password.Value,
-				ele.Global.V1.External.Opensearch.Ssl.RootCert.Value,
-				ele.Global.V1.External.Opensearch.Ssl.ServerName.Value,
-				ele.Global.V1.External.Opensearch.Auth.AwsOs.AccessKey.Value,
-				ele.Global.V1.External.Opensearch.Auth.AwsOs.SecretKey.Value,
-				ele.Global.V1.External.Opensearch.Backup.S3.Settings.RoleArn.Value,
-			)
+func (p *PullConfigsImpl) getOSpassword() (string, error) {
+	for _, ip := range p.infra.Outputs.AutomatePrivateIps.Value {
+		if stringutils.SliceContains(p.exceptionIps, ip) {
+			continue
 		}
+		p.sshUtil.getSSHConfig().hostIP = ip
+		rawOutput, err := p.sshUtil.connectAndExecuteCommandOnRemote(GET_OS_PASSWORD, true)
+		if err != nil {
+			return "", err
+		}
+		return strings.TrimSpace(rawOutput), nil
 	}
-	return nil
+	return "", nil
+
+}
+
+func (p *PullConfigsImpl) getExternalOpensearchDetails(a2ConfigMap map[string]*dc.AutomateConfig, dbType string) (*ExternalOpensearchToml, error) {
+	if dbType == TYPE_AWS {
+		for _, ele := range a2ConfigMap {
+			if ele.Global.V1.External.Opensearch != nil &&
+				ele.Global.V1.External.Opensearch.Auth != nil &&
+				ele.Global.V1.External.Opensearch.Auth.AwsOs != nil {
+				return setExternalOpensearchDetails(ele.Global.V1.External.Opensearch.Nodes[0].Value,
+					ele.Global.V1.External.Opensearch.Auth.AwsOs.Username.Value,
+					ele.Global.V1.External.Opensearch.Auth.AwsOs.Password.Value,
+					ele.Global.V1.External.Opensearch.Ssl.RootCert.Value,
+					ele.Global.V1.External.Opensearch.Ssl.ServerName.Value,
+					ele.Global.V1.External.Opensearch.Auth.AwsOs.AccessKey.Value,
+					ele.Global.V1.External.Opensearch.Auth.AwsOs.SecretKey.Value,
+					ele.Global.V1.External.Opensearch.Backup.S3.Settings.RoleArn.Value,
+				), nil
+			}
+		}
+	} else if dbType == TYPE_SELF_MANAGED {
+		osPass, err := p.getOSpassword()
+		if err != nil {
+			return nil, status.Wrap(err, status.ConfigError, "unable to fetch Opensearch password")
+		}
+		for _, ele := range a2ConfigMap {
+			if ele.Global.V1.External.Opensearch != nil &&
+				ele.Global.V1.External.Opensearch.Auth != nil &&
+				ele.Global.V1.External.Opensearch.Auth.BasicAuth != nil {
+				return setExternalOpensearchDetails(ele.Global.V1.External.Opensearch.Nodes[0].Value,
+					ele.Global.V1.External.Opensearch.Auth.BasicAuth.Username.Value,
+					osPass,
+					ele.Global.V1.External.Opensearch.Ssl.RootCert.Value,
+					ele.Global.V1.External.Opensearch.Ssl.ServerName.Value,
+					"",
+					"",
+					ele.Global.V1.External.Opensearch.Backup.S3.Settings.RoleArn.Value,
+				), nil
```

Comment:
Will moving getOsPassword() inside the IF block be better ?
```suggestion
		
		for _, ele := range a2ConfigMap {
			if ele.Global.V1.External.Opensearch != nil &&
				ele.Global.V1.External.Opensearch.Auth != nil &&
				ele.Global.V1.External.Opensearch.Auth.BasicAuth != nil {
				osPass, err := p.getOSpassword()
		                 if err != nil {
			                return nil, status.Wrap(err, status.ConfigError, "unable to fetch Opensearch password")
		                 }
				return setExternalOpensearchDetails(ele.Global.V1.External.Opensearch.Nodes[0].Value,
					ele.Global.V1.External.Opensearch.Auth.BasicAuth.Username.Value,
					osPass,
					ele.Global.V1.External.Opensearch.Ssl.RootCert.Value,
					ele.Global.V1.External.Opensearch.Ssl.ServerName.Value,
					"",
					"",
					ele.Global.V1.External.Opensearch.Backup.S3.Settings.RoleArn.Value,
				), nil
```

Reply:
yes its better, i have made the change